### PR TITLE
Automatically generate a "signature" when CMake is invoked

### PR DESCRIPTION
**Pull request by jlconlin**
This is a request that customers have made so that they know exactly which version of all the dependencies they have when it is built.

### CODE REVIEW

**Comment by jlconlin** [comment]
This is ready to go and adds ability to do `configure_file`.

**Comment by jlconlin** [comment]
Someone please review.

File: cmake.py [hunk 1]
```diff
@@ -49,6 +49,28 @@ def fetch_subprojects(state):
...
+    """
+    contents = textwrap.dedent(
+        """
+        if( NOT ROOT_DIRECTORY )
+            set( ROOT_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR} )
+        endif()
+        if ( NOT PYTHON_EXECUTABLE )
+            find_package( PythonInterp )
+            if ( NOT PYTHONINTERP_FOUND )
+                message( FATAL_ERROR "Python interpeter installation was not found." )
```
Comment (whaeck) [inline cmake.py]: There's a typo here: interpeter

File: cmake.py [hunk 2]
```diff
@@ -632,10 +667,12 @@ def generate():
     description.collect_subprojects(state, state['project path'])
     contents = \
         """
-cmake_minimum_required( VERSION 3.2 ) 
+cmake_minimum_required( VERSION 3.2 )
 set( CMAKE_CONFIGURATION_TYPES "Debug;Release" CACHE STRING "Supported configuration types" FORCE )
         """
     contents += fetch_subprojects(state)
+    if state["signature"]:
```
Comment (jlconlin) [inline cmake.py]: Use `if state.get("signature")` instead. This will avoid `KeyError` issues.

File: cmake.py [hunk 3]
```diff
@@ -49,6 +49,28 @@ def fetch_subprojects(state):
 
     return contents
 
+def signature():
+    """
+    Generate the 'signature' for the project
+    """
+    contents = textwrap.dedent(
```
Comment (jlconlin) [inline cmake.py:56]: Wrap this in a block that ensures it is not a subproject.

File: cmake.py [hunk 4]
```diff
@@ -389,51 +410,72 @@ def link_dependencies(state):
...
-    compile_flags=target_flags_expression(state)
-    link_flags=lto_flags_expression(state)
-    
+    sources = '"\n             "${CMAKE_CURRENT_SOURCE_DIR}/'.join(sources)
+
+    for conf in state['configure files']:
+        hpp = conf[:-3]
+        contents += textwrap.dedent("""
+            configure_file( "${{CMAKE_CURRENT_SOURCE_DIR}}/{conf}"
+                            "${{CMAKE_CURRENT_BINARY_DIR}}/{hpp}" )
```
Comment (jlconlin) [inline cmake.py]: Add `@ONLY` to ensure that it only replaces the @VARIABLE@ stuff.

File: signature.py [hunk 5]
```diff
@@ -8,33 +8,27 @@
...
-        for subproject in os.listdir( os.getcwd() ):
-            if os.path.isdir( os.path.join( os.getcwd(), subproject ) ):
-                os.chdir( subproject )
-                project_signature( git )
-                os.chdir( root )
-
-        os.chdir( '..' )
+    git[ project_name ] = project_signature()
+    root = os.getcwd()
+    for subproject in os.listdir( "subprojects" ):
```
Comment (jlconlin) [inline signature.py]: Add `try` block to catch if `subprojects` doesn't exist as subdirectory.

**Review by jlconlin — COMMENTED** [review]
asdfasd

**Comment by jlconlin** [comment]
Change description_generate to add configure file extensions

**Comment by jlconlin** [comment]
@apmccartney Add signature.json to install part of CMake.

**Review by whaeck — CHANGES_REQUESTED** [review]

**Review by apmccartney — APPROVED** [review]
Verified comments addressed

**Pull request merged**